File: euler29.py
# Сколько различных членов имеет последовательность a^b для
# 2 ≤ a ≤ 100 и 2 ≤ b ≤ 100?

from project_euler_defs import *

n = 100

# 1 вариант
start_time = datetime.now()
powers = []
for a in range(2, n + 1):
    for b in range(2, n + 1):
        if (x := a ** b) not in powers:
            powers.append(x)
print(len(powers))
print(datetime.now() - start_time)

# 2 вариант
start_time = datetime.now()
powers = set([a ** b for a in range(2, n + 1) for b in range(2, n + 1)])
print(len(powers))
print(datetime.now() - start_time)

File: go/euler29.go
// euler29 считает количество различных членов последовательности a^b
// для 2 ≤ a ≤ 100 и 2 ≤ b ≤ 100.
//
// Решение задачи Эйлера №29.
package main

import (
	"fmt"
	"time"
)

func euler29() {
	n := 100

	startTime := time.Now()
	powers := make(map[int]bool)
	for a := 2; a <= n; a++ {
		for b := 2; b <= n; b++ {
			pow := 1
			for i := 0; i < b; i++ {
				pow *= a
			}
			powers[pow] = true
		}
	}
	fmt.Println(len(powers))
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	powers = make(map[int]bool)
	for a := 2; a <= n; a++ {
		for b := 2; b <= n; b++ {
			pow := 1
			for i := 0; i < b; i++ {
				pow *= a
			}
			powers[pow] = true
		}
	}
	fmt.Println(len(powers))
	fmt.Println(time.Since(startTime))
}

func main() {
	euler29()
}
